### PR TITLE
[Draft]: Adds Reserved IP support.

### PR DESCRIPTION
**Pull request by haardikdharma10**
### Description of your changes

Adds Reserved IP support.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/civo/instance/v1alpha1/types.go [hunk 1]
```diff
@@ -40,6 +40,9 @@ type CivoInstanceConfig struct {
 
 	// +optional
 	PublicIPRequired string `json:"publicIPRequired,omitempty"`
+
+	// +optional
+	ReservedIP string `json:"reserved_ip,omitempty"`
```
Comment (RealHarshThakur) [inline apis/civo/instance/v1alpha1/types.go:45]:
```suggestion
       // Name of the CivoIP custom resource
	ReservedIP string `json:"reserved_ip,omitempty"`
```

File: internal/controller/civoinstance/civoinstance.go [hunk 2]
```diff
@@ -170,6 +171,25 @@ func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
...
+			ip, err := e.civoClient.GetIP(ip.Status.AtProvider.ID)
+			if err != nil {
+				return managed.ExternalCreation{}, err
+			}
+			err = e.civoClient.AssignIP(ip.ID, createInstance.Status.AtProvider.ID, "instance")
+			if err != nil {
+				return managed.ExternalCreation{}, err
+			}
+		} else {
+			return managed.ExternalCreation{}, err
```
Comment (RealHarshThakur) [inline internal/controller/civoinstance/civoinstance.go:189]: If it's empty, we have to unassign any reserved IP. Like we do in CCM

File: internal/controller/civoip/civoip.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+	cr.SetConditions(xpv1.Creating())
+
+	return managed.ExternalCreation{
+		ExternalNameAssigned: true,
+	}, nil
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+
```
Comment (RealHarshThakur) [inline internal/controller/civoip/civoip.go:153]: You should still get the object, update the status with latest information

File: internal/controller/civoip/civoip.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		return managed.ExternalObservation{}, errors.New(errNotCivoIP)
+	}
+	civoIP, err := e.civoClient.GetIP(cr.Status.AtProvider.ID)
+	if err != nil {
+		return managed.ExternalObservation{ResourceExists: false}, err
+	}
+	if civoIP == nil {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+	if civoIP.IP != "" {
```
Comment (RealHarshThakur) [inline internal/controller/civoip/civoip.go:121]: Status should be updated with latest information

File: pkg/civocli/cli.go [hunk 5]
```diff
@@ -259,3 +259,60 @@ func (c *CivoClient) DeleteK3sCluster(name string) error {
...
+		if strings.Contains(err.Error(), "DatabaseIPNotFoundError") {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return ip, nil
+}
+
+// ListIPs returns a list of civo IPs.
+func (c *CivoClient) ListIPs() (*civogo.PaginatedIPs, error) {
```
Comment (RealHarshThakur) [inline pkg/civocli/cli.go:287]: I don't think List, Assign,Unassign and Delete methods here are required. They feel like duplicating the calls from Civo Go client

File: e2e/ip_test.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	civoinstance "github.com/crossplane-contrib/provider-civo/apis/civo/instance/v1alpha1"
+	civoip "github.com/crossplane-contrib/provider-civo/apis/civo/ip/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestReservedIPBasic(t *testing.T) {
+
+	g := NewGomegaWithT(t)
+
+	ip := &civoip.CivoIP{
```
Comment (RealHarshThakur) [inline e2e/ip_test.go:20]: Can we have the IP creation in a separate method which checks if the IP already exists with the name and doesn't recreate it if not needed?

File: e2e/ip_test.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		return ip.Status.AtProvider.Address
+	}, "2m", "5s").ShouldNot(BeEmpty())
+
+	instance := &civoinstance.CivoInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "e2e-test-instance",
+		},
+		Spec: civoinstance.CivoInstanceSpec{
+			InstanceConfig: civoinstance.CivoInstanceConfig{
+				Hostname:  "e2e-test-instance",
```
Comment (RealHarshThakur) [inline e2e/ip_test.go:60]: We're supposed to pass in the name of the reserved IP here

File: e2e/ip_test.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+			},
+		},
+	}
+
+	fmt.Println("Creating Instance")
+	err = e2eTest.tenantClient.Create(context.TODO(), instance)
+	g.Expect(err).ShouldNot(HaveOccurred())
+
+	g.Eventually(func() string {
+		err = e2eTest.tenantClient.Get(context.TODO(), client.ObjectKeyFromObject(instance), instance)
```
Comment (RealHarshThakur) [inline e2e/ip_test.go:73]: We don't need this, this should be hard coded during instance creation

**Review by RealHarshThakur — CHANGES_REQUESTED** [review]

File: internal/controller/civoinstance/civoinstance.go [hunk 9]
```diff
@@ -129,6 +133,35 @@ func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
 		return managed.ExternalObservation{}, errors.Wrap(err, errGenObservation)
 	}
 
+	ip := &ipv1alpha1.CivoIP{}
+	err = e.kube.Get(ctx, types.NamespacedName{Name: civoInstance.ReservedIP}, ip)
```
Comment (RealHarshThakur) [inline internal/controller/civoinstance/civoinstance.go:137]: We should wrap this in an if block which checks if civoinstance.ReservedIP is not empty

Comment (RealHarshThakur) [reply internal/controller/civoinstance/civoinstance.go:137]: Can you also write an e2e test which creates an instance without reserved IP?

File: e2e/ip_test.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+	fmt.Println("Creating Instance")
+	err = e2eTest.tenantClient.Create(context.TODO(), instance)
+	g.Expect(err).ShouldNot(HaveOccurred())
+
+	g.Eventually(func() string {
+		err = e2eTest.tenantClient.Get(context.TODO(), client.ObjectKeyFromObject(ip), ip)
+		return ip.Status.AtProvider.AssignedTo.ID
+	}, "2m", "5s").Should(Equal(instance.Status.AtProvider.ID))
+}
+
```
Comment (RealHarshThakur) [inline e2e/ip_test.go:79]: Can we also have following tests please?
* Unassignment of reserved IP: removing the `ReservedIP` field from instance spec should unassign the reserved IP. 
* Deletion of reserved IP: Deleting the reserved IP after creation. Just to cover entire lifecycle.

File: examples/civo/instance/instance-nossh.yaml [hunk 11]
```diff
@@ -9,5 +9,6 @@ spec:
     size: g3.large
     hostname: myCrossplaneInstance2
     initialUser: crossplane
+    reserved_ip: test-crossplane-ip
```
Comment (RealHarshThakur) [inline examples/civo/instance/instance-nossh.yaml:12]: I think we should create a different example with a filename `instance-reservedIP` for this example

File: examples/civo/provider/provider.yaml [hunk 12]
```diff
@@ -5,7 +5,7 @@ metadata:
   name: civo-provider-secret
 type: Opaque
 data:
-  # credentials: <replace with base64 encoded Civo API KEY>
+  credentials: <replace with base64 encoded Civo API KEY>
```
Comment (RealHarshThakur) [inline examples/civo/provider/provider.yaml:8]: This should stay commented

File: examples/civo/instance/instance-nossh.yaml (hunk 11 above)
Comment (satakshigarg) [reply examples/civo/instance/instance-nossh.yaml:12]: @RealHarshThakur shall I rename this instance-nossh.yaml to instance-reservedIP or create a new example file ?

**Pull request closed without merging**